### ISSUE TITLE
Saving results to disk

### ISSUE THREAD

**Issue by crocha700** [rabernat/pyqg#9]
Guys, I just implemented a single-layer subclass (I'll merge my brach into develop soon). I've created a simple notebook using the model to reproduce a classical experiment by McWilliams JFM 1984. One thing that came up when I was doing this is that we may want to write the state variables to disk, so that we can study the evolution of the system.

Any thoughts on how to do the saving w/o compromising performance?

**Comment by francispoulin** [comment]
Hello Cesar,

Thanks. Did you post this somewhere on github?  I'd be happy to play with it if you wanted a tester.

I have no idea about how to save things quickly but as long as you save it only every so often maybe it wouldn't matter?  Maybe I'm very naive too.

Francis



------------------
Francis Poulin
Associate Professor
Associate Chair, Undergraduate Studies
Department of Applied Mathematics
University of Waterloo

email:           fpoulin@uwaterloo.ca
Web:            https://uwaterloo.ca/poulin-research-group/
Telephone:  +1 519 888 4567 x32637

________________________________

**Comment by crocha700** [comment]
Yes, I merged  it's merged with develop. Just update your fork and checkout develop.

**Comment by crocha700** [comment]
I'm picking this up. Do we agree that hdf5 is the best format to save parameters, diagnostics, and snapshots to disk?

**Comment by rabernat** [comment]
netcdf4, which uses hdf as a backend

what sort of api do you have in mind?

**Comment by crocha700** [comment]
I haven't thought of anything fancy. For the setup we can do something like:
~~~~{.python}
def _save_model_setup(self,):

    """Save setup  """

    fno = self.fno + 'setup.h5'

    _file_exist(self,fno)

    h5file = h5py.File(fno, 'w')
    
    h5file.create_dataset("grid/nx", data=(self.nx),dtype=int)
    h5file.create_dataset("grid/ny", data=(self.ny),dtype=int)
    h5file.create_dataset("grid/nz", data=(self.nz),dtype=int)
    h5file.create_dataset("grid/x", data=(self.x))
    h5file.create_dataset("grid/y", data=(self.y))
    h5file.create_dataset("grid/H", data=(self.Hi))
    h5file.create_dataset("grid/wv", data=self.wv)

    h5file.create_dataset("basestate/U", data=self.Ubg)
    h5file.create_dataset("basestate/V", data=self.Vbg)

    h5file.create_dataset("constants/beta", data=self.f)
    h5file.create_dataset("constants/f", data=self.beta)

   # and so on...

    h5file.close()
~~~~

For the snapshots I'm not sure yet...Ideas?

**Comment by rabernat** [comment]
By api, I mean the interface for the user, not the programmer.

How will I activate saving to disk? How will I specify which parameters to save, and the frequency of output?

Best to start with that and then figure out how to make it happen.

Please please use netcdf instead of hdf.

**Comment by crocha700** [comment]
So I thought of having a "save2disk" flag and a "tsave" in model.py.

**Comment by rabernat** [comment]
what about
```python
m = pyqg.model(..., write_netcdf_freq=3600, write_netcdf_fields=['qh', 'q'])
```

**Comment by crocha700** [comment]
Yeah, this seems great. But I'd say we have only 'qh' as default since all other fields can be easily obtained from qh...

**Comment by rabernat** [comment]
yes agreed
The documentation should clarified which fields are allowed in ``write_netcdf_fields``
